HID Switch to force sending of duplicate data and cleanup of JS API functions

JoergAtGithub opened the pull request:
I added a switch to force the sending of HID OutputReports with identical data, as needed by @Be-ing to support the S4 MK3 mapping: https://github.com/mixxxdj/mixxx/pull/4585#discussion_r813208966

After looking at the JavaScript API, I found out, that
`Q_INVOKABLE void send(const QList<int>& data, unsigned int length = 0) override`
is only there to comply with the inheritance of controller.c. But this function is practically useless. It doesn't send raw bytes as the siblings for MIDI and BULK, because this is a functionality, which the HID protocoll doesn't support. Instead it sends a OutputReport with hardcoded ReportID 0. I wonder, if this function was ever used anywhere.

Than there is
`Q_INVOKABLE void send(const QList<int>& data, unsigned int length, unsigned int reportID)`
which has already one unused argument (length).

Both send functions use QLIST as argument, while since #4521, the other functions of the JavaScript API use QByteArray.

I added a new API function with QByteArray
`Q_INVOKABLE void sendOutputReport(unsigned int reportID, const QByteArray& dataArray, resendUnchangedReport = false)`
and extended the old send function backward compatibility:
`Q_INVOKABLE void send(const QList<int>& dataList, unsigned int length, unsigned int reportID, resendUnchangedReport = false)`
Note: I found no way to make an overloaded function work, which support both QList or QByteArray. Independend of the type in JavaScript, always the same function was called. 

Furthermore I noticed, that the comments describing these JavaScript API functions are no longer apply since #4521, I rewrote them (in Doxygen syntax). Especially the returned array of getInputReport became incompatible to the incommingdata function. And this compatibility was the only reason to prepend the reportId to the data array.



Now the HID API functions are:

**Unchanged since 2.3:**
`Q_INVOKABLE void send(const QList<int>& dataList, unsigned int length = 0) override`
IMHO this function makes no sense, but I kept it for backward compatibility

**Extended since 2.3:**
`Q_INVOKABLE void send(const QList<int>& dataList, unsigned int length, quint8 reportID, bool resendUnchangedReport = false)`
Added optional argument resendUnchangedReport - Note that the Length argument is unused and it uses QList

**New in 2.4 and now unified in content and type of the data array:**
`Q_INVOKABLE void sendOutputReport(quint8 reportID, const QByteArray& dataArray, bool resendUnchangedReport = false)`
`Q_INVOKABLE QByteArray getInputReport(quint8 reportID)`
`Q_INVOKABLE void sendFeatureReport(quint8 reportID, const QByteArray& reportData)`
`Q_INVOKABLE QByteArray getFeatureReport(quint8 reportID)`

This is now similar to the basic IO functions in all other HID APIs I know - as well in naming as in order of arguments

JoergAtGithub committed:
Added new sendOutputReport wrappers to HidControllerJSProxy:
-With data as QList or QByteArray
-Without the unused length argument
-With a new argument to disable skipping of OutputReports with redundant data

JoergAtGithub committed:
Implemented skipIdenticalReports flag

JoergAtGithub committed:
Improved comments
Removed not working overloaded sendOutputReport with argument type QList

JoergAtGithub committed:
Added missing argument

JoergAtGithub committed:
Pre-commit fixes
Added missing sentence in comment

uklotzde commented on src/controllers/hid/hidcontroller.h:
```diff
@@ -60,21 +57,58 @@ class HidControllerJSProxy : public ControllerJSProxy {
+    /// @param skipIdenticalReports If set, the sending of the report is inhibited, when the data didn't changed
+    Q_INVOKABLE void sendOutputReport(const QByteArray& dataArray,
+            unsigned int reportID,
+            bool skipIdenticalReports = true) {
```
`resendUnchangedReport = false`

- "identity" is a different concept
- using `false` as the default value is more common
- `true` triggers an action -> positive logic
- only applies to a single report, not multiple

uklotzde commented on src/controllers/hid/hidiooutputreport.cpp:
```diff
@@ -80,7 +82,7 @@ bool HidIoOutputReport::sendCachedData(QMutex* pHidDeviceAndPollMutex,
     }
 
-    if (!m_lastSentData.compare(m_cachedData)) {
+    if (!m_skipIdenticalReports && !m_lastSentData.compare(m_cachedData)) {
```
This condition is wrong

Be-ing commented:
@JoergAtGithub could you take care of @uklotzde's review comments above?

I would prefer to not add yet another function to the API for this one niche use case.

JoergAtGithub committed:
Renamed skipIdenticalReports to resendUnchangedReport and inverted the logical meaning of the flag

JoergAtGithub committed:
Fixed wrong logical expression in if condition

JoergAtGithub committed:
Changed order of arguments for sendFeatureReport and sendOutputReport (both are introduced after 2.3).
ReportId is now first argument, as in all other HID APIs I know.

JoergAtGithub committed:
Changed type of reportId from unsigned int to const quint8
Passed by reference now

JoergAtGithub committed:
Changed type of reportId from unsigned int to const quint8 for more functions
Passed by reference now

JoergAtGithub committed:
Removed redundant ReportId from returned data-array of getInputReport:
-Now the arguments are identical with the similar function getFeatureReport
-This array is no longer compatible with incomingData anyway, because the type changed in #4521
-And even before, it was only compatible, under the assumption, that the HID device uses ReportIds
Note, that getInputReport is new for 2.4 and can be changed without breaking backward compatibility

Swiftb0y commented on src/controllers/hid/hidcontroller.h:
```diff
@@ -75,7 +75,7 @@ class HidControllerJSProxy : public ControllerJSProxy {
             unsigned int length,
             const quint8& reportID,
-            bool resendUnchangedReport = false) {
+            const bool& resendUnchangedReport = false) {
```
passing a cheap to copy value type by reference does not make much sense.

JoergAtGithub committed:
Changed small arguments, which are cheap to copy, from passing by reference to copy by value

JoergAtGithub committed:
Unified type for HID ReportId across the classes.

JoergAtGithub committed:
Unified argument order across the classes.

daschuer commented on src/controllers/hid/hidcontroller.h:
```diff
@@ -60,45 +57,79 @@ class HidControllerJSProxy : public ControllerJSProxy {
+    /// @param length Unused but mandatory argument
+    /// @param reportID 1...255 for HID devices that uses ReportIDs - or 0 for devices, which don't use ReportIDs
+    /// @param resendUnchangedReport If set, the report will also be send, if the data are unchanged since last sending
+    Q_INVOKABLE void send(const QList<int>& dataList,
```
You have changed the signature now from 

```
Q_INVOKABLE void send(const QList<int>& data, unsigned int length, unsigned int reportID) {
```

```
Q_INVOKABLE void send(const QList<int>& dataList, unsigned int length, quint8 reportID, bool esendUnchangedReport = false)
```

But the function is deprecated anyway. Can this cause any compatibility issues? If in doubt I would prefer no to touch it.

JoergAtGithub replied:
I tested it with the unmodified common-hid-packet-parser.js and it works flawless, with and without the optional argument.

daschuer replied:
Thank you for the double check!

daschuer commented on src/controllers/hid/hidcontroller.h:
```diff
@@ -60,45 +57,79 @@ class HidControllerJSProxy : public ControllerJSProxy {
+    ///        This function only works with HID devices, which don't use ReportIDs
+    /// @param dataList Data to send as list of bytes
+    /// @param length Unused optional argument
+    Q_INVOKABLE void send(const QList<int>& dataList, unsigned int length = 0) override {
```
How does this comment relate: 
https://github.com/mixxxdj/mixxx/blob/f2d670edeb483efb066bf7f00255d01df2f0017c/res/controllers/common-hid-packet-parser.js#L467

Introduced here: 
https://github.com/mixxxdj/mixxx/pull/1953

JoergAtGithub replied:
These are different tings:
There are two categories of HID devices:

1. Devices that use Report-IDs in no report (than the device can have only one report of each type Input/Output/Feature).
   hidapi expect the Report-ID byte for Output and Feature-Reports to be 0 in this case, but don't prepend this 0 to Input-Reports itself.

2. Devices that use a Report-ID in any report (allowed range of report IDs is 1...255).

Therefore you can use the sendOutputReport/sendFeatureReport functions and specify the Report-ID argument as 0, if you want send data to an HID device, that don't uses Report-IDs.

daschuer replied:
Ah OK, that makes sense.

daschuer commented:
Does it make sense to adjust res/controllers/common-hid-packet-parser.js to the new API?

JoergAtGithub committed:
Removed redundant this->

JoergAtGithub commented:
> Does it make sense to adjust res/controllers/common-hid-packet-parser.js to the new API?

In fact, this is what I had done for testing this PR. There is only one single line to change: https://github.com/mixxxdj/mixxx/blob/2c3b119d354c05da1c3caf1c11cb0937e415481a/res/controllers/common-hid-packet-parser.js#L694
this needs to be changed to:
`controller.sendOutputReport(this.reportId, Uint8Array.from(data).buffer);`

daschuer commented:
> In fact, this is what I had done for testing this PR. There is only one single line to change: 

For my understanding this saves some CPU cycles due to the QByteArray, right? Would you mind to add this commit? 

This PR looks good to me. 

@uklotzde: Can we merge?

JoergAtGithub commented:
> For my understanding this saves some CPU cycles due to the QByteArray, right? Would you mind to add this commit?

I don't know, what's the more efficient implementation. I would prefer to do it in a follow-up PR for common-hid-packet-parser.js

uklotzde reviewed (approved):
Thank you for this follow-up PR. Reordering the parameters to reflect the order within the actual data make sense. Not touching the Q_INVOKABLE functions goes without saying.

The data is copied twice into QByteArray? I didn't check if this behavior is unchanged from before.

LGTM
